### ISSUE TITLE
pulse penetration ratio calculation takes much more time than other features

### ISSUE THREAD

**Issue by komazsofi** [eEcoLiDAR/eEcoLiDAR#104]
Example 1.
------ Import is started ------
Number of points: 828317
Number of points in target: 828317
------ Computing neighborhood is started ------
build kd-tree: 10.313262 sec
------ Feature calculation is started ------
Feature "sigma_z" took 130.60 seconds
Feature "pulse_penetration_ratio" took 1336.95 seconds
Feature "slope" took 82.19 seconds
Feature "eigenv_3" took 114.05 seconds
Feature "echo_ratio" took 36.02 seconds
Feature "max_z" took 390.34 seconds
feature calc: 2090.169666 sec

Example 2.
------ Import is started ------
Number of points: 1969286
Number of points in target: 1969286
------ Computing neighborhood is started ------
build kd-tree: 28.908372 sec
------ Feature calculation is started ------
Feature "sigma_z" took 498.38 seconds
Feature "pulse_penetration_ratio" took 13991.64 seconds
Feature "slope" took 226.56 seconds
Feature "eigenv_3" took 434.94 seconds
Feature "echo_ratio" took 79.43 seconds
Feature "max_z" took 925.89 seconds
feature calc: 16156.876132 sec

Example 3.
------ Import is started ------
Number of points: 3427630
Number of points in target: 3427630
------ Computing neighborhood is started ------
build kd-tree: 51.410019 sec
------ Feature calculation is started ------
Feature "sigma_z" took 551.82 seconds
Feature "pulse_penetration_ratio" took 43054.16 seconds
Feature "slope" took 366.29 seconds
Feature "eigenv_3" took 615.07 seconds
Feature "echo_ratio" took 138.46 seconds
Feature "max_z" took 1585.47 seconds
feature calc: 46311.276115 sec

**Comment by romulogoncalves** [comment]
Would it be possible to provide a script to test this issue? Or share how it was executed, i.e., share the parameters?

**Comment by komazsofi** [comment]
Be aware that the code run before the implementation of tiling the data set internally for the process.

Code:
``` python
import argparse
import time
import numpy as np

parser = argparse.ArgumentParser()
parser.add_argument('path_of_laserchicken', help='The path of laserchicken')
parser.add_argument('input', help='absolute path of input point cloud')
parser.add_argument('target', help='target points as las file')
parser.add_argument('radius', help='radius of the volume')
parser.add_argument('output', help='absolute path of output point cloud')
args = parser.parse_args()

import sys
sys.path.insert(0, args.path_of_laserchicken)

from laserchicken import read_las
from laserchicken.keys import point
from laserchicken.volume_specification import InfiniteCylinder
from laserchicken.compute_neighbors import compute_neighborhoods
from laserchicken.feature_extractor import compute_features
from laserchicken.write_ply import write

print("------ Import is started ------")

pc = read_las.read(args.input)
target = read_las.read(args.target)

print(("Number of points: %s ") % (pc[point]['x']['data'].shape[0]))
print(("Number of points in target: %s ") % (target[point]['x']['data'].shape[0]))

print("------ Computing neighborhood is started ------")

start = time.time()
indices_cyl=compute_neighborhoods(pc, target, InfiniteCylinder(np.float(args.radius)))
end = time.time()
difftime=end - start
print(("build kd-tree: %f sec") % (difftime))

print("------ Feature calculation is started ------")

start1 = time.time()

compute_features(pc, indices_cyl, target, ['max_z','echo_ratio','eigenv_1', 'eigenv_2', 'eigenv_3',
'normal_vector_1','normal_vector_2','normal_vector_3','slope','pulse_penetration_ratio','sigma_z'], InfiniteCylinder(np.float(args.radius)))

end1 = time.time()
difftime1=end1 - start1
print(("feature calc: %f sec") % (difftime1))

write(target,args.output)
```

The script was run over all the las files in <our webdav path>/01_Work/zsofia/geobia/Results/Data/Lauw_island_tiles

using this script:
for f in <path>/*.las;do python computefea_wtargets_cylinder.py <path> $f $f

**Comment by romulogoncalves** [comment]
Fixed by #108
